merkle/objhasher/objhasher.go

```diff
@@ -0,0 +1,51 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package objhasher
+
+import (
+	"crypto/sha256" // Use SHA256 to match ObjectHash.
+
+	"github.com/benlaurie/objecthash/go/objecthash"
+)
+
+// ObjectHasher uses ObjectHash to compute leaf hashes.
+var ObjectHasher = objhasher{}
+
+// ObjectHash does not use `1` as any of its type prefixes,
+// preserving domain separation.
+const nodeHashPrefix = 1
+
+type objhasher struct{}
+
+// HashEmpty returns the hash of an empty element for the tree
+func (t objhasher) HashEmpty() []byte {
+	return sha256.New().Sum(nil)
+}
+
+// HashLeaf returns the object hash of leaf, which must be a JSON object.
+func (t objhasher) HashLeaf(leaf []byte) []byte {
+	hash := objecthash.CommonJSONHash(string(leaf))
```

Comment:
Thanks for the detailed explanation.

It sounds like an "object hash strategy" is the way to accommodate this. I won't bring the debate of overwriting hashes back to life, we don't have quorum here and it shouldn't block your PR. We're still lacking the "glue" between Tree setup and the hashes being used; when I (or someone else?) bridge that gap I'll talk to you if there are issues.